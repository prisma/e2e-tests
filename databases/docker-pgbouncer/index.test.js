const {
  client1WithoutQueryStringParamCall,
  client2WithoutQueryStringParamCall,
  client1WithQueryStringParamCall,
  client2WithQueryStringParamCall,
  client1WithoutQueryStringParam,
  client2WithoutQueryStringParam,
  pgBouncerMetaClient,
  client1WithQueryStringParam,
  client2WithQueryStringParam,
} = require('.')

describe('should test Prisma client and PgBouncer', () => {
  beforeAll(async () => {
    await pgBouncerMetaClient.connect()
  })

  afterAll(async () => {
    await client1WithoutQueryStringParam.$disconnect()
    await client2WithoutQueryStringParam.$disconnect()
    await client1WithQueryStringParam.$disconnect()
    await client2WithQueryStringParam.$disconnect()
    await pgBouncerMetaClient.end()
    return
  })

  it('should fail with docker pgbouncer without the pgbouncer query string param', async () => {
    let query0 = await pgBouncerMetaClient.query('SHOW VERSION')
    console.log(query0.rows)

    try {
      let query1 = await client1WithoutQueryStringParamCall()
      console.log({ query1 })

      /*
       * Query engines instances name prepared statements serially s0, s1 and so on. Without the `pgbouncer=true` flag,
       * prepared statements are not cleaned up in PgBouncer. By using different Clients, we expect the both instances to use s0
       * The conflicting client call should throw "prepared statement s0 already exists"
       */
      let query2 = await client2WithoutQueryStringParamCall()
      console.log({ query2 })

      expect(1).toEqual(0) // The code should never reach here
    } catch (e) {
      expect(e.toString()).toMatchSnapshot()
    }
  })

  it('should work with docker pgbouncer with the pgbouncer query string param', async () => {
    let query0 = await pgBouncerMetaClient.query('SHOW VERSION')
    console.log(query0.rows)

    const data = await client1WithQueryStringParamCall()
    expect(data).toMatchSnapshot()

    const data1 = await client2WithQueryStringParamCall()
    expect(data1).toMatchSnapshot()
  })
})
