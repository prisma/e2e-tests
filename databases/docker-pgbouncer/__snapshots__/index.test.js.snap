// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test Prisma client and PgBouncer should fail with docker pgbouncer without the pgbouncer query string param 1`] = `
"Error: 
Invalid \`prisma.user.findMany()\` invocation:


  Error occurred during query execution:
ConnectorError(ConnectorError { user_facing_error: None, kind: QueryError(Error { kind: Db, cause: Some(DbError { severity: \\"ERROR\\", parsed_severity: Some(Error), code: SqlState(\\"26000\\"), message: \\"prepared statement \\\\\\"s0\\\\\\" does not exist\\", detail: None, hint: None, position: None, where_: None, schema: None, table: None, column: None, datatype: None, constraint: None, file: Some(\\"prepare.c\\"), line: Some(505), routine: Some(\\"FetchPreparedStatement\\") }) }) })"
`;

exports[`should test Prisma client and PgBouncer should work with docker pgbouncer with the pgbouncer query string param 1`] = `
Array [
  Object {
    "email": "alice-d97227bb-97b6-4a6d-81bf-04322b686165@prisma.io",
    "id": "d97227bb-97b6-4a6d-81bf-04322b686165",
    "name": "Alice",
  },
]
`;

exports[`should test Prisma client and PgBouncer should work with docker pgbouncer with the pgbouncer query string param 2`] = `
Array [
  Object {
    "email": "alice-d97227bb-97b6-4a6d-81bf-04322b686165@prisma.io",
    "id": "d97227bb-97b6-4a6d-81bf-04322b686165",
    "name": "Alice",
  },
]
`;
